confuse.py

```diff
@@ -134,36 +135,193 @@ def __init__(self, filename, reason=None):
+
+    try:
+        return wraps(func)(inner)
+    except AttributeError:
+        # in v2 they don't ignore missing attributes
+        # v3: https://github.com/python/cpython/blob/3.8/Lib/functools.py
+        # v2: https://github.com/python/cpython/blob/2.7/Lib/functools.py
+        inner.__name__ = func.__name__
+        return inner
+
+
+def all_subclasses(cls):
+    return set(cls.__subclasses__()).union(
+        s for c in cls.__subclasses__() for s in all_subclasses(c))
+
+
 class ConfigSource(dict):
-    """A dictionary augmented with metadata about the source of the
+    '''A dictionary augmented with metadata about the source of the
     configuration.
-    """
-    def __init__(self, value, filename=None, default=False):
-        super(ConfigSource, self).__init__(value)
+    '''
+    def __init__(self, value=UNSET, filename=None, default=False):
+        # track whether a config source has been set yet
+        self.loaded = value is not UNSET
+        super(ConfigSource, self).__init__(value if self.loaded else {})
         if filename is not None and not isinstance(filename, BASESTRING):
             raise TypeError(u'filename must be a string or None')
         self.filename = filename
         self.default = default
 
     def __repr__(self):
-        return 'ConfigSource({0!r}, {1!r}, {2!r})'.format(
-            super(ConfigSource, self),
-            self.filename,
-            self.default,
-        )
+        return '{}({}, filename={}, default={})'.format(
+            self.__class__.__name__,
+            dict.__repr__(self)
+            if self.loaded else '[Unloaded]'
+            if self.exists else "[Source doesn't exist]",
+            self.filename, self.default)
+
+    @property
+    def exists(self):
+        '''Does this config have access to usable configuration values?'''
+        return self.loaded or self.filename and os.path.isfile(self.filename)
+
+    def load(self):
+        '''Ensure that the source is loaded.'''
+        if not self.loaded:
+            self.config_dir()
+            self.loaded = self._load() is not False
+        return self
+
+    def _load(self):
+        '''Load config from source and update self.
+        If it doesn't load, return False to keep it marked as unloaded.
+        Otherwise it will be assumed to be loaded.
+        '''
+
+    def config_dir(self, create=True):
+        '''Create the config dir, if there's a filename associated with the
+        source.'''
+        if self.filename:
+            dirname = os.path.dirname(self.filename)
+            if create and dirname and not os.path.isdir(dirname):
+                os.makedirs(dirname)
+            return dirname
+        return None
 
     @classmethod
-    def of(cls, value):
-        """Given either a dictionary or a `ConfigSource` object, return
-        a `ConfigSource` object. This lets a function accept either type
-        of object as an argument.
-        """
+    def is_of_type(cls, value):
```

Comment:
This mechanism for automatically detecting which type of source to use is a little too "magical," I think, for a first cut. If it's OK with you, can we leave this out for now? (Concretely, an issue with searching for all subclasses of a given class is that it is implicitly extensible—anyone who creates a subclass of `ConfigSource` will automatically/transparently be extending this overload system, which can be somewhat surprising.)

Perhaps as a more convenient alternative to explicitly using `YamlSource` we can just hard-code some API to use this that class exclusively?

Reply:
Yeah I agree with you - if we want to be more explicit, then we can just do the same as was done with `as_template`.

Currently, by default any class extending `ConfigSource` would have to extend `is_of_type` in order to work with automatic detection so there isn't as much risk. My biggest concern is that resolution order is (?) undefined (?) or at least not clear. Tbh it was just an easy way to try to have ConfigSource be extensible without modifying the confuse core, but I agree that it isn't optimal and I'm fine moving to a more explicit definition.